### ISSUE TITLE
Unexpected EOF while looking for matching `"' when copying the secret (JSON format) to a file

### ISSUE THREAD

**Issue by StephenDsouza90** [actions/runner#1974]
**Describe the bug**

I have credentials that I want to store in a file called configuration.json. The credentials are in JSON format and are stored in the secrets. One of the fields in JSON is with the name "password" that includes a single quote (example: **g1fWhME'}8vS**). In one of the steps in the Github actions CI/CD pipeline, I copy the secret to this file.

When copying, the following error occurs: **unexpected EOF while looking for matching `"'**

**To Reproduce**

1. Copy and paste the below JSON format into a secret

```json
{
    "applicationMode": "production",
    "production": {
        "database": {
            "dbname": "dbname",
            "host": "host",
            "port": "port",
            "user": "user",
            "password": "g1fWhME'}8vS"
        }
    }
}
```

2. Create a simple CI/CD pipeline

```
name: Build

on: [push, pull_request]

jobs:
  configuration:
    name: configuration
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repository
        uses: actions/checkout@v2

      - name: Copy secrets to configuration file
        run: |
          echo '${{ secrets.CONFIGURATION }}' > configuration.json

```

**Expected behavior**

The expected behavior is that Github actions will treat the value of the password as a string and copy the JSON formatted credentials into the file in a proper JSON format.

**Solution currently used**
```
      - name: Copy secrets to configuration file
        run: |
          cat > configuration.json <<EOF
          ${{ secrets.CONFIGURATION }}
          EOF

```
## Runner Version and Platform
Not applicable here

## What's not working?
<img width="992" alt="Screenshot 2022-06-27 at 9 09 46 PM" src="https://user-images.githubusercontent.com/53117931/176017397-8505acbe-049f-4094-81aa-7c7b3b11ab1b.png">

Error message: unexpected EOF while looking for matching `"'

## Job Log Output
```
Run echo '***
  echo '***
      ***
      ***
          ***
              ***
              ***
              ***
              ***
              ***
          ***
      ***
  ***' > configuration.json
  shell: /usr/bin/bash -e ***0***
  /home/runner/work/_temp/53a756dc-c384-4732-8743-79c45d7d5a5d.sh: line 9: unexpected EOF while looking for matching `"'
  Error: Process completed with exit code 2.
```

## Runner and Worker's Diagnostic Logs
Not applicable here

**Comment by nikola-jokic** [comment]
Hi @StephenDsouza90,

The issue here is that, when the template is evaluated, it will replace the content inside the `${{}}` with the correct value. After the evaluation, steps will be applied. You can think of that place where the evaluation happens as the placeholder to which the evaluation will expand your input. That way, having the `'` character when replaced will break the command, causing this issue.

This is actually by design, and I hope this explanation is helpful. That is actually the reason your solution you provided is working. 

I am closing this issue now, but feel free to comment on it :relaxed:

**Comment by jameswilliamknight** [comment]
tl;dr for anyone else reading this

- injecting variables into a workflow gets done first
- result gets executed second

so escape any quotes you inject that clash with the surrounding ones.

Escape double quotes:
```yml
- name: Copy secrets to configuration file
  run: |
    ESCAPED_CONFIGURATION=$(echo "${{ secrets.CONFIGURATION  }}" | sed 's/"/\\"/g')
    echo "${{ secrets.CONFIGURATION }}" > configuration.json
#        ^                            ^                                   ^
```

Escape single quotes:
```yml
- name: Copy secrets to configuration file
  run: |
    ESCAPED_CONFIGURATION=$(echo "${{ secrets.CONFIGURATION }}" | sed "s/'/\\\'/g")
    echo '$ESCAPED_CONFIGURATION' > configuration.json
#        ^                      ^                                        ^
```

**Comment by lippoliv** [comment]
I sumbled across this with "`" (backtick) today. 

```yaml
      - name: Upload to INT
        shell: bash
        working-directory: ${{ vars.PLUGIN_FOLDER_NAME }}
        run: |
          export RCLONE_FTP_PASS=`rclone obscure "${{ secrets.INT_FTP_PASSWORD }}"`
```

I was able to "fix" it (I mean it's not particulary wrong as @nikola-jokic explained) by using an ENV


```yaml
      - name: Upload to INT
        shell: bash
        working-directory: ${{ vars.PLUGIN_FOLDER_NAME }}
        env:
          FTP_PASSWORD: ${{ secrets.INT_FTP_PASSWORD }}
        run: |
          export RCLONE_FTP_PASS=$(rclone obscure "$FTP_PASSWORD")
```

(I also use `$()` to wrap the command but I think it's not that relevant)

The reason why it doesn't brake this way is that the env var is evaluated when the comment runs, while `${{}}` would be replaced befor command execution and then the command is "corrupt".


Maybe this is a good alternative for others as well :)
